runtime/gc_vlhgc/WriteOnceCompactor.cpp

```diff
@@ -1782,9 +1782,14 @@ MM_WriteOnceCompactor::fixupRoots(MM_EnvironmentVLHGC *env)
 					}
 				} else { 
 					/* Only system/app classloaders can have a null classloader object (only during early bootstrap) */
-					Assert_MM_true((classLoader == _javaVM->systemClassLoader)
-							|| (classLoader == _javaVM->applicationClassLoader)
-							|| (classLoader == _javaVM->extensionClassLoader));
+#if defined(J9VM_OPT_SNAPSHOTS)
+					if (!IS_RESTORE_RUN(_javaVM))
+#endif /* defined(J9VM_OPT_SNAPSHOTS) */
+					{
+						Assert_MM_true((classLoader == _javaVM->systemClassLoader)
+								|| (classLoader == _javaVM->applicationClassLoader)
+								|| (classLoader == _javaVM->extensionClassLoader));
+					}
```

Comment:
Frozen classes should be applicable to class loaders can not be unloaded. Otherwise it create interference with class unloading process. Current contract allows class loader object to be NULL for such class loaders, and there is the list how GC sees it. New contract (with isolation of frozen classes from GC) should follow the same rules. You can extend explicit list of class loaders if you like.
I think just removing this assertion is incorrect.